lib/shared/robot/README.md

```diff
@@ -0,0 +1,123 @@
+The "robot" mechanism
+=====================
+
+As discussed in `lib/child-writer`, communication between the main Etcher
+application and its elevated writer process happens through an IPC (Inter
+Process Communication) channel. In a nutshell, we emit every single line that
+the writer process prints to the parent as a "message". Since these "lines"
+need to convey non-trivial information such as progress information, speed,
+final computer checksums, etc we are in need of a basic form of a "protocol".
+
+The "robot" module is the entity that implements this protocol, and provides
+utility functions to read/send messages using the protocol targetted at both
+parties (the client and the writer processes).
+
+For example:
+
+*Child process:*
+
+```js
+robot.printMessage('my-message-type', {
+  my: {
+    message: 'data'
+  }
+});
+```
+
+*Parent process:*
+
+```js
+const message = robot.parseMessage(line);
+
+console.log(robot.getCommand(message));
+> 'my-message-type'
+
+console.log(robot.getData(message));
+> {
+>   my: {
+>     message: 'data'
+>   }
+> }
+```
+
+The codename "robot" is inspired by [xz][xz-man], which provides a `--robot`
+option that makes the tool print machine-parseable output:
+
+```
+--robot
+    Print messages in	a machine-parsable format.  This  is  intended
+    to  ease	writing	 frontends  that  want	to  use	 xz instead of
+    liblzma, which may be the	case with various scripts.  The	output
+    with  this  option  enabled  is  meant  to  be  stable across xz
+    releases.	 See the section ROBOT MODE for	details.
+```
+
+To enable the "robot" option, we standardised the presence of an
+`ETCHER_CLI_ROBOT` environment variable. You can check if the mode is enabled
+by using the `.isEnabled()` static function that the robot module provides:
+
+```js
+if (robot.isEnabled()) {
+  console.log('The robot option is enabled');
+}
+```
+
+The current protocol that we use is based on JSON. The writer process
+stringifies a JSON object as a string, and prints it. The client then gets the
+line, parses it as JSON, and accesses the object.
+
+For example, the writer process may have a ficticious internal object that
+looks like this:
+
+```js
+{
+  percentage: 50,
+  stage: 'validation'
+}
+```
+
+That object can be stringified as `{"percentage":50,"stage":"validation"}` and
+printed to `stdout`.
+
+The contents and structure of this messages is what the "robot" module is
+mainly concerned with. Each "message" consists of a type (a "command" in robot
+parlance) and and arbitrary data object.
+
+This is what a valid robot message looks like:
+
+```json
+{
+  "command": "progress",
+  "data": {
+    "percentage": 50
+  }
+}
+```
```

Comment:
What do you think about describing the general data structure before showing method usage? I think that would give more context to the usage examples.